### PR TITLE
Some CI -related tweaks

### LINKED ISSUE #263: Support OTP 24

### PR DESCRIPTION
**Pull request by paulo-ferraz-oliveira**
- [x] add CI cache
- [x] remove unused Elvis rule
- [x] make `dialyzer` runs Ok on the first try (no need to `|| true`)
- [x] use `erl_anno` where appropriate

### CODE REVIEW

File: .github/workflows/erlang.yml [hunk 1]
```diff
@@ -10,17 +10,23 @@ jobs:
 
     strategy:
       matrix:
-        otp: [21.3, 22.3, 23.3]
+        otp: [21, 22, 23, 24]
```
Comment (paulo-ferraz-oliveira) [inline .github/workflows/erlang.yml]: `setup-beam` will choose the highest possible version in the <x> range (just check the https://github.com/AdRoll/rebar3_format/pull/275/checks).

File: .github/workflows/erlang.yml [hunk 2]
```diff
@@ -10,17 +10,23 @@ jobs:
...
 
     steps:
     - uses: actions/checkout@v2
     - uses: erlef/setup-beam@v1
       with:
         otp-version: ${{matrix.otp}}
-        rebar3-version: '3.15.2'
-    - name: Compile
-      run: rebar3 compile
```
Comment (paulo-ferraz-oliveira) [inline .github/workflows/erlang.yml:22]: Automatically done by `rebar3`.

File: .github/workflows/erlang.yml [hunk 3]
```diff
@@ -10,17 +10,23 @@ jobs:
...
-    - name: Generate dialyzer PLT
-      run: rebar3 dialyzer || true # to generate the PLT :(
```
Comment (paulo-ferraz-oliveira) [inline .github/workflows/erlang.yml:24]: No longer required as per changes to `rebar.config`.

File: .github/workflows/erlang.yml [hunk 4]
```diff
@@ -10,17 +10,23 @@ jobs:
...
+        rebar3-version: '3.16'
+    - name: Restore _build
```
Comment (paulo-ferraz-oliveira) [inline .github/workflows/erlang.yml:21]: Yay, cache 🎉

File: elvis.config [hunk 5]
```diff
@@ -2,5 +2,4 @@
   [{config,
     #{dirs => ["src/**"],
       filter => "*.erl",
-      ruleset => erl_files,
-      rules => [{elvis_style, no_tabs, disable}]}}]}].
```
Comment (paulo-ferraz-oliveira) [inline elvis.config:6]: No longer used.

File: rebar.config [hunk 6]
```diff
@@ -17,17 +17,9 @@
...
-
-{dialyzer,
- [{warnings, [no_return, unmatched_returns, error_handling, underspecs]},
-  {get_warnings, true},
-  {plt_apps, top_level_deps},
-  {plt_extra_apps, []},
-  {plt_location, local},
-  {base_plt_apps, [erts, kernel, stdlib, runtime_tools, syntax_tools]},
-  {base_plt_location, "_build/global_plt"}]}.
+{project_plugins, [rebar3_hex, rebar3_lint, rebar3_hank]}.
```
Comment (paulo-ferraz-oliveira) [inline rebar.config:20]: I bumped all these to latest to (hopefully) ease maintenance; hope that's Ok.

File: rebar.config [hunk 7]
```diff
@@ -17,17 +17,9 @@
...
-{project_plugins,
- [{rebar3_hex, "~> 6.11.6"}, {rebar3_lint, "~> 0.4.1"}, {rebar3_hank, "~> 1.1.3"}]}.
-
-{dialyzer,
- [{warnings, [no_return, unmatched_returns, error_handling, underspecs]},
-  {get_warnings, true},
-  {plt_apps, top_level_deps},
-  {plt_extra_apps, []},
-  {plt_location, local},
-  {base_plt_apps, [erts, kernel, stdlib, runtime_tools, syntax_tools]},
```
Comment (paulo-ferraz-oliveira) [inline rebar.config:29]: `runtime_tools, syntax_tools` moved to `.app.src`; the rest is not really required.

**Comment by paulo-ferraz-oliveira** [comment]
So, here we are: https://github.com/AdRoll/rebar3_format/pull/275/checks?check_run_id=3192825934#step:6:69. Without my changes this fails on OTP 24 too, but I didn't dig into it.

File: rebar.config (hunk 6 above)
Comment (elbrujohalcon) [reply rebar.config:20]: Nope. That's against NextRoll's policy. We like our builds to be _predictable_ and we use https://github.com/AdRoll/rebar3_depup to stay up-to-date, instead.

**Review by elbrujohalcon — CHANGES_REQUESTED** [review]

**Comment by elbrujohalcon** [comment]
Regarding the tests… it looks like the precedence of `catch` changed slightly in OTP24 and it doesn't require parentheses around it anymore…
That's odd.

**Comment by elbrujohalcon** [comment]
I would personally remove OTP24 support from this PR… Merge everything else and then create a different PR just to add OTP24 support.
In that PR, I would include in `otp_formatter` the changes that the OTP team included in `erl_prettypr` between OTP23 and OTP24.

**Comment by paulo-ferraz-oliveira** [comment]
- [ ] compare erl_prettypr from 23 to 24…
- [ ] apply the same patches to otp_formatter
- [ ] if it’s worth it, add new tests for the new stuff
- [ ] then apply the “worthy” patches to default_formatter

File: rebar.config (hunk 6 above)
Comment (paulo-ferraz-oliveira) [reply rebar.config:20]: Cool. Reverting.

Comment (elbrujohalcon) [reply rebar.config:20]: I already reverted… oops! I did it incorrectly 🤦🏻

Comment (elbrujohalcon) [reply rebar.config]: Now fixed for good.

Comment (paulo-ferraz-oliveira) [reply rebar.config]: 150a4a9

File: src/formatters/otp_formatter.erl [hunk 8]
```diff
@@ -1194,12 +1190,23 @@ lay_clause_expressions([H | T], Ctxt) ->
     end;
 lay_clause_expressions([], _) -> empty().
 
+is_last_and_before_empty_line(H, [], #ctxt{empty_lines = EmptyLines}) ->
```
Comment (paulo-ferraz-oliveira) [inline src/formatters/otp_formatter.erl:1193]: This block is what's making `otp_formatter_SUITE` fail at the moment.

Comment (elbrujohalcon) [reply src/formatters/otp_formatter.erl:1193]: Yeah. I _guess_ it's because of the additional line at the end of the file… but I'm not 100% sure.

**Comment by paulo-ferraz-oliveira** [comment]
As per @elbrujohalcon's recommendation I'm turning this into a "merely maintenance" pull request. A new one will be opened for the OTP 24, stuff.

File: .github/workflows/erlang.yml (hunk 1 above)
Comment (paulo-ferraz-oliveira) [reply .github/workflows/erlang.yml]: Comment makes no sense any more.

File: rebar.config (hunk 6 above)
Comment (paulo-ferraz-oliveira) [reply rebar.config]: It seems we were changing it at the same time 😄

File: src/formatters/otp_formatter.erl (hunk 8 above)
Comment (paulo-ferraz-oliveira) [reply src/formatters/otp_formatter.erl:1193]: This comment makes no sense anymore.

**Comment by paulo-ferraz-oliveira** [comment]
@elbrujohalcon, this has been re-purposed and is now ready for re-review.

**Review by elbrujohalcon — APPROVED** [review]

**Pull request merged**